modules/references/references.go

```diff
@@ -273,12 +277,19 @@ func findAllIssueReferencesBytes(content []byte, links []string) []*rawReference
 			if len(parts) != 5 || parts[0] != "" {
 				continue
 			}
-			if parts[3] != "issues" && parts[3] != "pulls" {
+			var sep string
+			if parts[3] == "issues" {
+				// Issues can be external
+				sep = "#"
+			} else if parts[3] == "pulls" {
+				// PRs are always local
+				sep = "!"
```

Comment:
So this will break reference pull request via `!`?

Reply:
I don't see the problem as `!` will be an allowed way to reference pulls. But if someone want to use only `#` for referencing issues and pulls I think there are two possible solutions:
 - Gitea only use the `!` when auto-referencing PRs if the repository referenced have external tracker.
 - Make it a setting of the pull request unit of each repo if pulls are prefered to be referenced using `!`. I would prefer this but you probably need the first point anyhow.

But that would make the code more complex since you need to check settings of the referenced pull request. Just using `!` in any case is simpler.

Reply:
> 
> 
> So this will break reference pull request via `!`?

@lunny  The `#` syntax will still be perfectly valid for both PRs and issues when they are hosted in Gitea; the only difference is when issues are numeric and externally tracked: `#` can be used for issues and `!` will force a local link, which happens to be a PR. Otherwise the user would not be able to link the PR easily.